### PR TITLE
fix(#31): handle user consent cancellation (#24)

### LINKED ISSUE #31: NullPointerException when user refuses consent
Hi,

**Version used** : 1.2.0 (with KC 19.0.3, wildfly legacy)

When a user refuses Apple Consent screen (to share its email) by clicking the cancel button : 

Then we got this NPE : 

```
{
  "class": "org.keycloak.services.managers.AuthenticationSessionManager",
  "method": "lambda$getUserSession$7",
  "line": 220
},
{
  "class": "org.keycloak.utils.LockObjectsForModification",
  "method": "lockObjectsForModification",
  "line": 75
},
{
  "class": "org.keycloak.utils.LockObjectsForModification",
  "method": "lockUserSessionsForModification",
  "line": 65
},
{
  "class": "org.keycloak.services.managers.AuthenticationSessionManager",
  "method": "getUserSession",
  "line": 220
},
{
  "class": "org.keycloak.services.resources.IdentityBrokerService",
  "method": "checkAccountManagementFailedLinking",
  "line": 1072
},
{
  "class": "org.keycloak.services.resources.IdentityBrokerService",
  "method": "error",
  "line": 860
},
{
  "class": "at.klausbetz.provider.AppleIdentityProviderEndpoint",
  "method": "authResponse",
  "line": 81
},
{
  "class": "jdk.internal.reflect.NativeMethodAccessorImpl",
  "method": "invoke0"
},
[...]
```

I think that this functional case should be handled like ACCESS_DENIED. But in both [callback.cancelled()](https://github.com/keycloak/keycloak/blob/19.0.3/services/src/main/java/org/keycloak/services/resources/IdentityBrokerService.java#L848) or [callback.error()](https://github.com/keycloak/keycloak/blob/19.0.3/services/src/main/java/org/keycloak/services/resources/IdentityBrokerService.java#L860) authenticationSession is required in [checkAccountManagementFailedLinking](https://github.com/keycloak/keycloak/blob/19.0.3/services/src/main/java/org/keycloak/services/resources/IdentityBrokerService.java#L1071) method. 

Before submitting a PR, what do you think about fetching authenticationSession here before this line : 

https://github.com/klausbetz/apple-identity-provider-keycloak/blob/09a2721e425bcfe54eba298cb2925c857608fd1c/src/main/java/at/klausbetz/provider/AppleIdentityProviderEndpoint.java#L65

```
    @POST
    public Response authResponse(@FormParam(OAUTH2_PARAMETER_STATE) String state, @FormParam(OAUTH2_PARAMETER_CODE) String authorizationCode, @FormParam(OAUTH2_PARAMETER_USER) String user, @FormParam(OAuth2Constants.ERROR) String error) {
        if (state == null) {
            event.event(EventType.IDENTITY_PROVIDER_LOGIN);
            event.error(Errors.IDENTITY_PROVIDER_LOGIN_FAILURE);
            return ErrorPage.error(session, null, Response.Status.BAD_GATEWAY, Messages.IDENTITY_PROVIDER_UNEXPECTED_ERROR);
        }

        IdentityBrokerState idpState = IdentityBrokerState.encoded(state);
        String clientId = idpState.getClientId();
        String tabId = idpState.getTabId();

        AuthenticationSessionModel authSession = this.callback.getAndVerifyAuthenticationSession(state);
        session.getContext().setAuthenticationSession(authSession);

[...]
```

just like it is done in [AbstractOAuth2IdentityProvider#authResponse()](https://github.com/keycloak/keycloak/blob/19.0.3/services/src/main/java/org/keycloak/broker/oidc/AbstractOAuth2IdentityProvider.java#L485)  ?

It fixes the issue. I think [this code](https://github.com/klausbetz/apple-identity-provider-keycloak/blob/09a2721e425bcfe54eba298cb2925c857608fd1c/src/main/java/at/klausbetz/provider/AppleIdentityProviderEndpoint.java#L82-L83) can be removed too 

What do you think @klausbetz ?

### PR DESCRIPTION
**Pull request by daviddelannoy**
Hi @klausbetz 

I propose this PR to fix #31 

- handling user consent decline by cancelling the whole workflow and redirecting him/her to login page (fixing NPE)
- 2 minors changes : 
    - checking `state` parameter like it is done in [AbstractOAuth2IdentityProvider](https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/broker/oidc/AbstractOAuth2IdentityProvider.java#L481)
    - returning an IDENTITY_PROVIDER_LOGIN event instead of a LOGIN event when error occurs. You may have changed it on purpose so let me know, but it also seems closer to what is done in the `AbstractOAuth2IdentityProvider`.

I tested it successfully with both Keycloak 20.0.1 & 20.0.3

PS. : I did not make non-regression tests on token exchange endpoint

### CODE REVIEW

File: src/main/java/at/klausbetz/provider/AppleIdentityProviderEndpoint.java [hunk 1]
```diff
@@ -58,19 +57,24 @@ public AppleIdentityProviderEndpoint(AppleIdentityProvider appleIdentityProvider
...
-            event.error(Errors.IDENTITY_PROVIDER_LOGIN_FAILURE);
-            return ErrorPage.error(session, null, Response.Status.BAD_REQUEST, Messages.INVALID_REQUEST);
+            return errorIdentityProviderLogin(Messages.INVALID_REQUEST, Response.Status.BAD_REQUEST);
         }
+
+        AuthenticationSessionModel authSession = this.callback.getAndVerifyAuthenticationSession(state);
+        session.getContext().setAuthenticationSession(authSession);
+
         if (error != null) {
             logger.error(error + " for broker login " + appleIdentityProvider.getConfig().getProviderId());
```
Comment (klausbetz) [inline src/main/java/at/klausbetz/provider/AppleIdentityProviderEndpoint.java:76]: Could you please lower the log level here to debug?  
Sure, I did this back then. However, it's no fatal error and a warn should still make clear something went wrong.

File: src/main/java/at/klausbetz/provider/AppleIdentityProviderEndpoint.java [hunk 2]
```diff
@@ -58,19 +57,24 @@ public AppleIdentityProviderEndpoint(AppleIdentityProvider appleIdentityProvider
...
-            if (error.equals(ACCESS_DENIED)) {
+            if (error.equals(ACCESS_DENIED) || error.equals(USER_CANCELLED_AUTHORIZE)) {
```
Comment (klausbetz) [inline src/main/java/at/klausbetz/provider/AppleIdentityProviderEndpoint.java:77]: Great handling 👍

**Comment by klausbetz** [comment]
Token-exchange is still fine 👍

File: src/main/java/at/klausbetz/provider/AppleIdentityProviderEndpoint.java (hunk 1 above)
Comment (daviddelannoy) [reply src/main/java/at/klausbetz/provider/AppleIdentityProviderEndpoint.java:76]: yes sure

File: src/main/java/at/klausbetz/provider/AppleIdentityProviderEndpoint.java (hunk 2 above)
Comment (daviddelannoy) [reply src/main/java/at/klausbetz/provider/AppleIdentityProviderEndpoint.java:77]: :)

**Review by klausbetz — APPROVED** [review]

**Pull request merged**